whole-tale/gwvolman
# Update dataone tales

Fixes issue #107. This is a rewrite of a much uglier PR (#111 ). This PR does _not_ include the refactored unit tests, but _does_ give a little more clarity into the Tale updating changes.

The first two commits just pull things into a few classes to make things easier down the road. The next few fix up a few pep issues. The last commit fixed this issue where one except statement was swallowing an exception from a nested try/catch. I flattened the try/catch blocks and things seem to be working fine.

### Reviewer Note

#115 should be merged after this PR is merged.

### Updating

A Tale is 'updated' when it's published a second time. If a Tale is published to DataONE, and then a copy is made and 
 subsequently published the `datacite:IsDerivedFrom` relation pointing to the first Tale's doi is added to the resource map of the newly published object (see example below for clarity). 

When a Tale is published to DataONE two times, the previous system metadata for the EML is retrieved. A new system metadata document is made from it, but with values such for md5 and size updated. This system metadata document is uploaded with the new EML document, and effectively signals DataONE to obsolete the package.

## Testing

1. Create a Tale
2. Publish it to DataONE
3. View the landing page for that Tale in a new tab (don't close the tab!)
4. Publish again
5. Refresh the URL from step 3, you should see a notification that there is an updated version available
_

1. Create a Tale
2. Publish it to DataONE
3. Make the Tale public
4. Sign out and sign in under a new account
5. Copy the Tale from step 1
6. Publish it to DataONE
7. Copy the ID of the resource map (the uuid in the picture below)
8. Visit https://dev.nceas.ucsb.edu/knb/d1/mn/v2/object/<RESOURCE_MAP_ID>
9. Search for "datacite"
10. Check that it references the copied Tale's doi


Example:

I published a Tale [here](https://dev.nceas.ucsb.edu/view/urn:uuid:12539245-06fc-482e-bec4-da2db24490c2), copied it and republished [here](https://dev.nceas.ucsb.edu/view/urn:uuid:e3cb4097-4182-4615-a8c4-74c91cb8a928). In the _new_ resource map found [here]( https://dev.nceas.ucsb.edu/knb/d1/mn/v2/object/urn:uuid:e3cb4097-4182-4615-a8c4-74c91cb8a928), you can see that the datacite:IsDerivedFrom points to the DOI of the first package.
